docs/specs/src/hwp_4_formal.md

```diff
@@ -0,0 +1,534 @@
+
+Like most DHT architectures, Holochain uses a "nearness" algorithm to compute the "distance" distance between the 256-bit Blake2b basis hash of the data or metadata to be stored and the 256-bit Ed25519 public key (network address) of nodes. Basically, it is the responsibility of the nodes *nearest* a basis hash to store data and metadata for it.
+
+Holochain's validating, graphing, gossiping DHT implementation is called **rrDHT**.
+
+rrDHT is designed with a few performance requirements/characteristics in mind.
+
+1. It must have a compact and computationally simple representational model for identifying which nodes are responsible for which content, and which nodes actually hold which content. (A **"world model"** of what is where.)
+2. It must have **lookup speeds** at least as fast as Kademlia's binary trees ( $\mathcal{O}(n \log n)$ ). Current testing shows an average of 3 hops/queries to reach an authority with the data.
+3. It must be **adjustable** to be both resilient and performant across many DHT compositional make-ups (reliability of nodes, different network topologies, high/low usage volumes, etc.)
+
+**World Model:** The network location space is a circle comprising of the range of unsigned 32-bit numbers, in which the location $0$ is adjacent to the location ${2^{32}}-1$.
+
+$$
+\begin{aligned}
+L &: \{l_0,\dots,l_{2^{32}-1}\} \\
+l_{2^{32}-1} &< l_0
+\end{aligned}
+$$
+
+The larger 256-bit address space of the DHT, consisting of 256-bit "basis hashes" $B$ (Blake2b-256 hashes of addressable content $C$, and Ed25519 public keys of agents $K$), is mapped to the smaller network location space via a function $\mathsf{map\_to\_loc}(b)$, which is the XOR of 8 × 32-bit segments of the hashes. The original address is still kept with the data or metadata, so collisions in the smaller space are not a concern.
+
+$$
+\begin{aligned}
+E &: \{e_0,\dots,e_{2^{256}-1}\} \\
+K &: \{k_0,\dots,k_{2^{256}-1}\} \\
+C &: E \cup K \\
+\mathsf{hash} &: C \rightarrow B \\
+\mathsf{map\_to\_loc} &: B \rightarrow L \\
+|L| &\approx |B| \cdot 2^8
+\end{aligned}
+$$
+
+Each agent has a network location $l_k$ in this 32-bit space as well as an arc size $s_{arc}$ indicating how large an 'arc' of the location circle they are claiming authority for. The storage arc $ARC_{l_k}$ defines the range of basis hashes for which a node claims authority. The arc spreads clockwise from $l_k$.
+
+$$
+ARC_{l_k} : \{l_k,\dots,l_k + s_{arc}\}
+$$
+
+This guarantees that a node can rapidly resolve any basis hash $b$ to the most likely candidate for an authority $k_{best}$ by comparing the basis hash's network location $l_b$ to all the authorities they know about in the set $L_K$ using the following algorithm, in pseudocode:
+
+```
+k_best = L_K
+    .sort_ascending()
+    .filter(l_k -> l_k <= l_b)
+    .last()
+```
```

Comment:
just curious, does holochain currently use this algorithm? the "most likely" guarantee only holds under the assumption that all arcs are the same size, which is mostly true (the only exceptions are zero and full arcs), but may be even less true in the future

Reply:
With the addition of filtering out arcs that don't cover the location, this looks good. I thought you were sorting to find the agent closest to the basis, assuming that that would be the most likely arc to cover it, and that's the part that would only be true if all arcs are the same size (if arcs are different sizes, a closer agent could not cover it and a futher agent still could).

Reply:
correction #2: I realised it doesn't need to filter by arc; that wouldn't permit the remote peer to suggest a more likely candidate. Also didn't realise Rust doesn't naturally wraparound `u32` arithmetic, so you'd need to introduce some modulo there: `.sort_ascending_by(l_p -> (l_b - l_p) mod 2^32)`

I'm editing the text to say that you can ask a peer if they know of a more likely candidate than they, if they're the closest you know about. But I don't know -- do peers normally have knowledge of peers to the left of them by virtue of having received connection requests from them? they're not technically expected to hold peers to the left if they're not authorities for them, right?

Reply:
Right, you only know about peers to the left because they initiated gossip with you. I'm not sure if we hold onto agent info for those peers though.